### PR TITLE
v1 of a bisulfite sequencing pipeline

### PR DESCRIPTION
**Pull request by chrisamiller**

### CODE REVIEW

File: bisulfite/index_cram.cwl [hunk 1]
```diff
@@ -0,0 +1,21 @@
+#!/usr/bin/env cwl-runner
+
+cwlVersion: v1.0
+class: CommandLineTool
+label: "samtools index"
+arguments: [
+    "cp", $(inputs.cram.path), "$(runtime.outdir)/$(inputs.cram.basename)",
+    { valueFrom: " && ", shellQuote: false },
+    "/opt/samtools/bin/samtools", "index", $(inputs.cram.path), "$(runtime.outdir)/$(inputs.cram.basename).bai"
```
Comment (tmooney) [inline bisulfite/index_cram.cwl]: This should be `.crai` instead of `.bai`.

File: bisulfite/pileup.cwl [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+    - class: ResourceRequirement
+      ramMin: 16000
+      coresMin: 4
+arguments: [
+    { valueFrom: "-q", position: -10 },
+    { valueFrom: $(runtime.cores), position: -9 },
+    { valueFrom: "-w", position: -8 },
+    { valueFrom: "pileup_stats.txt", position: -7 },
+    { shellQuote: false, valueFrom: "|" },
+    "/bin/gzip"
```
Comment (tmooney) [inline bisulfite/pileup.cwl]: Should this call `bgzip` instead, in case we want to tabix index the VCF later?

File: bisulfite/pileup.cwl [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+            prefix: "-i"
+            position: -2
+    reference: 
+        type: string
+        inputBinding:
+            prefix: "-r"
+            position: -1
+outputs:
+    vcf:
+        type: File
```
Comment (tmooney) [inline bisulfite/pileup.cwl]: This could use `type: stdout` to refer to the file named by the `stdout:` above.

File: bisulfite/pileup.cwl [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+    - class: ResourceRequirement
+      ramMin: 16000
+      coresMin: 4
+arguments: [
+    { valueFrom: "-q", position: -10 },
+    { valueFrom: $(runtime.cores), position: -9 },
+    { valueFrom: "-w", position: -8 },
+    { valueFrom: "pileup_stats.txt", position: -7 },
+    { shellQuote: false, valueFrom: "|" },
+    "/bin/bgzip"
```
Comment (tmooney) [inline bisulfite/pileup.cwl]: In the current version of `mgibio/bisulfite`, `bgzip` is located at `/opt/htslib/bin/bgzip`.

File: bisulfite/bam_to_trimmed_fastq_and_biscuit_alignments.cwl [hunk 5]
```diff
@@ -0,0 +1,60 @@
...
+        out:
+            [fastq1, fastq2]
+    biscuit_align:
+        run: biscuit_align.cwl
+        in:
+            reference_index: reference_index
+            fastq1: 
+                source: trim_fastq/fastq1
+            fastq2: 
+                source: trim_fastq/fastq2
```
Comment (tmooney) [inline bisulfite/bam_to_trimmed_fastq_and_biscuit_alignments.cwl]: What's the significance of having these explicitly list a `source`?

File: bisulfite/vcf2bed.cwl [hunk 6]
```diff
@@ -0,0 +1,30 @@
...
+
+cwlVersion: v1.0
+class: CommandLineTool
+label: "Biscuit vcf2bed wrapper"
+baseCommand: ["/usr/bin/bsvcf2bed"]
+requirements:
+    - class: ShellCommandRequirement
+    - class: ResourceRequirement
+      ramMin: 16000
+      coresMin:
```
Comment (tmooney) [inline bisulfite/vcf2bed.cwl]: `coresMin` here is missing a value.

File: bisulfite/bam_to_trimmed_fastq_and_biscuit_alignments.cwl [hunk 7]
```diff
@@ -0,0 +1,60 @@
+#!/usr/bin/env cwl-runner
+
+cwlVersion: v1.0
+class: Workflow
+label: "bam to trimmed fastqs and biscuit alignments"
+requirements:
+    - class: MultipleInputFeatureRequirement
+    - class: SubworkflowFeatureRequirement
+    - class: InlineJavascriptRequirement
```
Comment (tmooney) [inline bisulfite/bam_to_trimmed_fastq_and_biscuit_alignments.cwl]: It doesn't look like this currently uses the `InlineJavascriptRequirement` or the `MultipleInputFeatureRequirement`.

File: bisulfite/vcf2bed.cwl [hunk 8]
```diff
@@ -0,0 +1,30 @@
+#!/usr/bin/env cwl-runner
+
+cwlVersion: v1.0
+class: CommandLineTool
+label: "Biscuit vcf2bed wrapper"
+baseCommand: ["/usr/bin/bsvcf2bed"]
+requirements:
+    - class: ShellCommandRequirement
```
Comment (tmooney) [inline bisulfite/vcf2bed.cwl]: It doesn't look like this is currently doing anything that would require the `ShellCommandRequirement`.

File: bisulfite/pileup.cwl (hunk 4 above)
Comment (chrisamiller) [reply bisulfite/pileup.cwl]: yeah, just found that during my test run - push coming shortly

**Comment by chrisamiller** [comment]
new push fixes those comments and runs successfully on the small test data

**Pull request merged**